# Purpose:
-> To create another AI model, now, more complex. Content explained on the previous notebook will not be covered here.
Topic: Classification

About the used Data: I have imported them from the Keras, it has a lot of data about flowers. They are not pictures, they are just numeric information. At the end, my model will be able to identify the Specie of the flower.

In [3]:
# Importing:
import numpy as np
import pandas as pd
from IPython.display import clear_output as co
import matplotlib.pyplot as plt
import tensorflow as tf
from six.moves import urllib
import inspect
import tensorflow.compat.v2.feature_column as fc

if not hasattr(inspect, 'getargspec'):
    inspect.getargspec = inspect.getfullargspec

In [4]:
# Creating constants for supproting later
COLUMN_NAMES = ['SepalLength', 'SepalWidth', 'PetalLength', 'PetalWidth', 'Species']
SPECIES = ['Setosa', 'Versicolor', 'Virginica']

In [5]:
# Installing datasets paths
trainPath = tf.keras.utils.get_file("iris_training.csv", "https://storage.googleapis.com/download.tensorflow.org/data/iris_training.csv")
testPath = tf.keras.utils.get_file("iris_test.csv", "https://storage.googleapis.com/download.tensorflow.org/data/iris_test.csv")

# Creating Pandas DataFrame
train_df = pd.read_csv(trainPath, names=COLUMN_NAMES, header=0) # Creating the DataFrame with the columns names defined earlier. The parameter  
                                                                # header = 0 means that the previous parameter should be at the first line #
                                                                # (header)

test_df = pd.read_csv(testPath, names=COLUMN_NAMES, header=0)

Lets see our data!

In [6]:
train_df.head()

SepalLength  SepalWidth  PetalLength  PetalWidth  Species
0          6.4         2.8          5.6         2.2        2
1          5.0         2.3          3.3         1.0        1
2          4.9         2.5          4.5         1.7        2
3          4.9         3.1          1.5         0.1        0
4          5.7         3.8          1.7         0.3        0

In [7]:
test_df.head()

SepalLength  SepalWidth  PetalLength  PetalWidth  Species
0          5.9         3.0          4.2         1.5        1
1          6.9         3.1          5.4         2.1        2
2          5.1         3.3          1.7         0.5        0
3          6.0         3.4          4.5         1.6        1
4          5.5         2.5          4.0         1.3        1

In [8]:
# Create the labels (y-axis)
train_y = train_df.pop('Species')
test_y = test_df.pop('Species')

COLUMN_NAMES.remove("Species") # This column has been moved to others vars

Observe that the values has been pop out

In [9]:
display(train_df)

SepalLength  SepalWidth  PetalLength  PetalWidth
0            6.4         2.8          5.6         2.2
1            5.0         2.3          3.3         1.0
2            4.9         2.5          4.5         1.7
3            4.9         3.1          1.5         0.1
4            5.7         3.8          1.7         0.3
..           ...         ...          ...         ...
115          5.5         2.6          4.4         1.2
116          5.7         3.0          4.2         1.2
117          4.4         2.9          1.4         0.2
118          4.8         3.0          1.4         0.1
119          5.5         2.4          3.7         1.0

[120 rows x 4 columns]

# More about the dataframe:
train_df: 120 entries with 4 features
test_df: 29 entries with 4 features

In [10]:
# Creating the Input Function
def input_fun(features, labels, training=True, batch_size=256):
    # Converting the inputs to a DataSet
    dataset = tf.data.Dataset.from_tensor_slices((dict(features), labels))

    # Shuffle and reapeat if you are in training mode.
    if training:
        dataset = dataset.shuffle(1000).repeat()
    
    return dataset.batch(batch_size)

# Observe: We did not use Epochs.

In [11]:
# Defining the features column
FEATURES_COLUMN = []
for key in train_df.keys():
    FEATURES_COLUMN.append(tf.feature_column.numeric_column(key=key))

Instructions for updating:
Use Keras preprocessing layers instead, either directly or via the `tf.keras.utils.FeatureSpace` utility. Each of `tf.feature_column.*` has a functional equivalent in `tf.keras.layers` for feature preprocessing when training a Keras model.


# Building the model
Since we are working with classification, there is a variety of different models we can use, such as:

DNNClassifier (Deep Neural Network)
Linear Classifier

I am going to use de DNN because there may be non-linear relations in our DatSet.

In [12]:
classifier = tf.estimator.DNNClassifier(
    feature_columns=FEATURES_COLUMN, # The features columns of our DataSet
    hidden_units=[30,10], # 2 Hidden Layers of 30 and 10 nodes (Neurons), for 3 Hidden Layers should be something: [amountNodes ,          
                        # amountNodes2, amountNodes3]
    n_classes=3 # The amount of clases the model will choose between
)

Instructions for updating:
Use tf.keras instead.
Instructions for updating:
Use tf.keras instead.
Instructions for updating:
Use tf.keras instead.
Instructions for updating:
Use tf.keras instead.
INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_model_dir': 'C:\\Users\\Eduardo\\AppData\\Local\\Temp\\tmp84c6jdxk', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_device_fn': None, '_protocol': None, '_eval_distribute': None, '_experimental_distribute': None, '_experimental_max_worker_delay_secs': None, '_session_creation_timeout_secs': 7200, '_checkpoint_save_graph_def': True, '_service': None, '_cluster_spec': ClusterSpec({}), '_task_type': 'worker', '

In [13]:
#Training!
classifier.train(
    input_fn= lambda: input_fun(train_df, train_y, training=True), # Did this because we did not defind an inner function at input_fun
    steps=5000 # The times the model will update the weights and biases afet a batch
)

Instructions for updating:
Use tf.keras instead.
INFO:tensorflow:Calling model_fn.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Instructions for updating:
Use tf.keras instead.
INFO:tensorflow:Done calling model_fn.
Instructions for updating:
Use tf.keras instead.
Instructions for updating:
Use tf.keras instead.
Instructions for updating:
Use tf.keras instead.
Instructions for updating:
Use tf.keras instead.
INFO:tensorflow:Create CheckpointSaverHook.
Instructions for updating:
Use tf.keras instead.
Instructions for updating:
Use tf.keras instead.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 0...
INFO:tensorflow:Saving checkpoints for 0 into C:\Users\Eduardo\AppData\Local\Temp\tmp84c6jdxk\model.ckpt.
INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 0..

In [14]:
# Checking output:
result = classifier.evaluate(input_fn=lambda: input_fun(test_df, test_y, training=False))
print("\033[32mAccuracy: " + str(result['accuracy']*100) + "%\033[0m")

INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2025-07-17T15:59:34
Instructions for updating:
Use tf.keras instead.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from C:\Users\Eduardo\AppData\Local\Temp\tmp84c6jdxk\model.ckpt-5000
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Inference Time : 0.53091s
INFO:tensorflow:Finished evaluation at 2025-07-17-15:59:34
INFO:tensorflow:Saving dict for global step 5000: accuracy = 0.96666664, average_loss = 0.3453255, global_step = 5000, loss = 0.3453255
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 5000: C:\Users\Eduardo\AppData\Local\Temp\tmp84c6jdxk\model.ckpt-5000
Accuracy: 96.66666388511658%


In [23]:
# Predicting according to system.input:
def input_pred_fn(features, batch_size=256):
    # Convert the inputs to a Dataset without labels.
    return tf.data.Dataset.from_tensor_slices(dict(features)).batch(batch_size)

predict = {}

for feature in COLUMN_NAMES:
    value = input(feature + ": ")
    predict[feature] = [float(value)]

predictions = classifier.predict(input_fn=lambda: input_pred_fn(predict))
for pred_dict in predictions:
    class_id = pred_dict['class_ids'][0]
    probability = pred_dict['probabilities'][class_id]

    print('\033[32mPrediction is "{}" ({:.1f}%)\033[32m'.format(
        SPECIES[class_id], 100 * probability))



INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from C:\Users\Eduardo\AppData\Local\Temp\tmp84c6jdxk\model.ckpt-5000
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
Prediction is "Setosa" (74.1%)
